### ISSUE TITLE
ra_syntax reports a false error in where predicate

### ISSUE THREAD

**Issue by regiontog** [rust-analyzer/rust-analyzer#584]
[This](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f1fe210f601495743af57290436f9bd8) playground demonstrates a case that ra_syntax fails to parse properly. I get [this](https://gist.github.com/regiontog/0e9b5f850eef1375336f0984a30b596f#file-gistfile1-txt-L51) syntax tree.

**Comment by matklad** [comment]
This the part of the parser which deals with where clauses:

https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ra_syntax/src/grammar/type_params.rs#L93-L156

Looks like we only handle path types, but other types should be handled as well. It makes sense to check what rust parser in libsyntax is doing:

https://github.com/rust-lang/rust/blob/7164a9f151a56316a382d8bc2b15ccf373e129ca/src/libsyntax/parse/parser.rs#L5421

**Comment by regiontog** [comment]
It seems like the two cases is lifetimes and types. I don't see any obvious restrictions i libsyntax, however neither placeholder types nor impl trait types should be allowed in where predicates AFAIK. The `types::type_with_bounds_cond` already can parse types so I'd rather avoid duplicating functionality. I'm thinking something like this as a parameter to the type gammar.

```rust 
struct TypeParameters {
    allow_bounds: bool,
    allow_impl_trait_type: bool,
    allow_placeholder_type: bool,
}

impl Default for TypeParameters {
    fn default() -> Self {
        TypeParameters {
            allow_bounds: true,
            allow_impl_trait_type: true,
            allow_placeholder_type: true,
        }
    }
}
```

It seems suprising to me though that the already available parameter `allow_bounds` is not passed down to e.g. `types::paren_or_tuple_type` given the recursion here. `types::paren_or_tuple_type` in calls `types::type_` which in turn calls `types::type_with_bounds_cond` with `allow_bounds` hardcoded to false. 

https://github.com/rust-analyzer/rust-analyzer/blob/0d2cb60f93c5e73c0171e83fef36b696cc5dd4ff/crates/ra_syntax/src/grammar/types.rs#L18-L37

**Comment by regiontog** [comment]
> if you find some other cases where rustc and rust-analyzer disagree, feel free to fix them as well.

I think there is a couple more in my project, I will look into it tomorrow I think.
